### ISSUE TITLE
Unable to migrate existing database to use plain text header in iOS

### ISSUE THREAD

**Issue by fabiorodella** [sqlcipher/sqlcipher#352]
I have recently been bitten by the `0xdead10cc` issue, and found https://github.com/sqlcipher/sqlcipher/issues/255. If I'm creating a new database and setting `cipher_plaintext_header_size` and `cipher_salt`, everything works great and I'm able to keep accessing the database later as normal as long as I'm passing the same salt value.

However, I'll need to migrate existing users, so I'm following the guidance in that issue to extract the salt from an existing database and migrate it to plain text header, but it doesn't seem to be working for me. At first I was using GRDB to do this but I tried using the SQLCipher API directly and I'm getting the same issue, which is error 26 as soon as I try to do anything else with the database (like `PRAGMA user_version = 1`). All of the code in my project is Swift.

This is the code I use to extract the salt from the database file:
```
extension Data {
    func hexadecimalString() -> String {
        let characters = (self as NSData).bytes.bindMemory(to: CChar.self, capacity: self.count)
        var hexString = ""

        for i in 0..<self.count {
            hexString += String(format: "%02.2hhx", arguments: [characters[i]])
        }

        return hexString
    }
}

let databaseData = try! Data(contentsOf: sharedDatabaseUrl, options: .alwaysMapped)
let saltData = databaseData.subdata(in: databaseData.startIndex..<databaseData.startIndex.advanced(by: 16))
let saltString = saltData.hexadecimalString()
```
(I have confirmed that the value that I get from this is the same as the one from `PRAGMA cipher_salt`)

And this is my test code to migrate the database:
```
var sqliteConnection: OpaquePointer?
var code: Int32 = 0

code = sqlite3_open_v2(path, &sqliteConnection, SQLITE_OPEN_CREATE | SQLITE_OPEN_READWRITE, nil)
guard code == SQLITE_OK else {
    fatalError("Opening database")
}

// encryptionKey in this context is a 26-character string
let data = encryptionKey.data(using: .utf8)!
code = data.withUnsafeBytes {
    sqlite3_key(sqliteConnection, $0.baseAddress, Int32($0.count))
}
guard code == SQLITE_OK else {
    fatalError("Keyeing database")
}

code = sqlite3_exec(sqliteConnection, "PRAGMA cipher_plaintext_header_size = 32;", nil, nil, nil)
guard code == SQLITE_OK else {
    fatalError("Setting plaintext header size")
}

code = sqlite3_exec(sqliteConnection, "PRAGMA cipher_salt = \"x'04accefd8e544c6a86b8ad71fa3b61b6'\";", nil, nil, nil)
guard code == SQLITE_OK else {
    fatalError("Setting plaintext header size")
}

code = sqlite3_exec(sqliteConnection, "PRAGMA user_version = 1;", nil, nil, nil)
guard code == SQLITE_OK else {
    fatalError("Setting user version to write header")
}
```

Everything goes fine until I get to the final statement, where the received error code is 26.

Is there anything obviously wrong with my code that I'm seeing, or any steps I'm missing? I also tried to play around with the order of PRAGMA statements, but it doesn't seem to change anything.

It's worth noting that my existing database is not in WAL mode, but I'm still seeing the `0xdead10cc` crashes.

SQLCipher version: 4.3.0

**Comment by fabiorodella** [comment]
It looks like the issue goes away when I put the database in WAL mode (which I had forgotten to do). Is it expected that `cipher_plaintext_header_size` and `cipher_salt` only work with WAL? If so, feel free to close this issue.

**Comment by sjlombardo** [comment]
Hello @fabiorodella - I believe the problem here is the logic in your migration code.  While it is necessary to set `cipher_salt` and `cipher_plaintext_header_size` immediately after keying to open a file with a plain text header, that is different then the case for an existing database that is being migrated. In that case the salt is readable and the header is *not* plaintext at that point. The proper process, when not using a raw key, would be something like this.

To start with the header is encrypted, prepended by the salt.

```
$ hexdump -C test.db | head -1
00000000  5d 53 6b 82 f4 23 fa 04  00 22 21 39 80 44 0f be  |]Sk..#..."!9.D..|
```

So, open it like a regular database, then after accessing the file, change the `cipher_plaintext_header_size` and force the header write.

```
sqlite> pragma key = 'test';
ok
sqlite> select count(*) from sqlite_master; -- trigger header read, currently it is encrypted
1
sqlite> PRAGMA cipher_plaintext_header_size = 32;
sqlite> PRAGMA user_version = 1; -- force header write
sqlite> .q
```

At this point he header is plaintext:

```
$ hexdump -C test.db | head -1
00000000  53 51 4c 69 74 65 20 66  6f 72 6d 61 74 20 33 00  |SQLite format 3.|
```

So, now, to open the file once it has a plaintext header, do the following:

```
sqlite> pragma key = 'test';
ok
sqlite> pragma cipher_plaintext_header_size = 32;
sqlite> pragma cipher_salt = "x'5d536b82f423fa040022213980440fbe'";
sqlite> .schema
CREATE TABLE t1(a,b);
```

**Comment by fabiorodella** [comment]
Interesting, so at first I just removed the `cipher_salt` setter during the migration pass, and kept getting error 26 on the call to `PRAGMA user_version = 1;`. I then looked more closely at your example and noticed that you had a `select count(*) from sqlite_master;` query before setting the plaintext header size, so I added that and it started working great!

Did I overlook some document that mentions the need for this? In any case, thank you very much for your help!
